server/social.coffee

```diff
@@ -385,24 +385,29 @@ module.exports = exports = (log, loga, argv) ->
     # see http://ward.asia.wiki.org/login-to-view.html
 
     if argv.restricted?
-
       allowedToView = (req) ->
-        allowed = []
         if argv.allowed_domains?
-          if Array.isArray(argv.allowed_domains)
-            allowed = argv.allowed_domains
-          else
-            # accommodate copy bug to be fixed soon
-            # https://github.com/fedwiki/wiki/blob/4c6eee69e78c1ba3f3fc8d61f4450f70afb78f10/farm.coffee#L98-L103
-            for k, v of argv.allowed_domains
-              allowed.push v
-        # emails = [ { value: 'ward.cunningham@gmail.com', type: 'account' } ]
-        emails = req.session?.passport?.user?.google?.emails
-        return false unless emails
-        for entry in emails
-          have = entry.value.split('@')[1]
-          for want in allowed
-            return true if want == have
+          try
+            allowed_domains = argv.allowed_domains
+            emails = req.session.passport.user.google.emails
+            for entry in emails
+              have = entry.value.split('@')[1]
+              for want in allowed_domains
+                return true if want == have
+          catch error
+            console.log "argv.allowed_domains exists, but there was an error. Make sure it's value is an array in your config."
+        if argv.allowed_usernames?
+          try
+            allowed_usernames = argv.allowed_usernames
+            idProvider = _.head(_.keys(req.session.passport.user))
+            switch idProvider
+              when 'github', 'twitter', 'oauth2'
+                username = req.session.passport.user[idProvider].username
```

Comment:
Not all auth providers provide a username, and even when they do it is not immutable. The only thing that is immutable, and unique, is the `id`.

Reply:
do you mean providers beyond the 4 that we currently account for in this plugin? (google, twitter, github, oauth2) Other than Google, which is dealt with separately in the `allowed_domains` case, the other 3 providers all provide a username, which would be unique within their system... (maybe you are worried about cases where a farm sets up multiple ID providers at the same time, and then there would be the potential for username collisions?)

Is it a big issue that the usernames aren't immutable? (if someone changed their username, the config would need to be changed to match) I kind of like that the `owner.json` files aren't filled with any secret or unknown information in this case (like with some internal ID).  Its just each person's public username, so we can do interesting things like composing `owner.json` files ourselves with known information, rather than only being able to generate them by claiming a site the typical way and having the IDProvider fill in the unknown information like their internal unique ID.

Reply:
The single prior use case for login to view which created `allowed_domains` just happened to use google. There is no reason why this is something that is restricted to Google as an authentication provider.

Authentication is done using the unique ID that the identity provider provides - and is stored in `owner.json`. All the other values there are simply there as that is the blob of data that the identity provider returns when the wiki is claimed.

Reply:
not sure what other providers also provide an email in the token, but the use of Google was kind of hard-coded into the existing code here on line 400.
https://github.com/fedwiki/wiki-security-passportjs/blob/99e25947f9b6dbe32c2af9f9ec9e30d07b420f5d/server/social.coffee#L400

Reply:
I'll just switch to using id, instead of username then.